### PR TITLE
Modify SEL to check endpoints against protocol rather than value_kind

### PR DESCRIPTION
**Pull request by hhan-schrodinger**
## Description
Added features to generic_data_parser.R script to work with the new endpoint manager when experiments are loaded. 

When an experiment is loaded that will create a new protocol, added logic that will: 
- Save the same "data column order" information currently being saved to Experiment via "createColumnOrderStates" into the new Protocol created by "createNewProtocol"
- Save incoming column names into the "data column"/"column name" codetable / data dictionary
- Create a value for "Strict Endpoint Matching" when a new protocol is created via experiment (based on config file)

When an experiment is loaded against an existing protocol that has strict endpoint matching enabled, added logic that will: 
- Instead of validating column headers against the "value_kind" table, validate them against the "data column order" lsStates of the protocol.
- Implement the warning & error messages that indicate the experiment columns that do not match the existing protocol's endpoints

## How Has This Been Tested?
- Loaded experiments to create new protocols and saw that the new column names/units/data types have been saved as new endpoints in the endpoint manager
- Loaded experiments against existing protocols with strict endpoint matching enabled to test that they threw errors indicating when the experiment column did not match the protocol's endpoints, and that they passed when they did match

Example of Error: 
Screen Shot 2022-11-04 at 12 07 07 PM

### CODE REVIEW

File: modules/GenericDataParser/src/server/generic_data_parser.R [hunk 1]
```diff
@@ -223,6 +223,77 @@ validateMetaData <- function(metaData, configList, username, formatSettings = li
 
   return(list(validatedMetaData=validatedMetaData, duplicateExperimentNamesAllowed=duplicateExperimentNamesAllowed, useExisting=useExisting))
 }
+
+
+saveIncomingEndpointData <- function(codes, codeKind) {
```
Comment (bffrost) [inline modules/GenericDataParser/src/server/generic_data_parser.R:228]: nit: this function name is pretty broad. Something narrower like `saveEndpointDDictValues` or `saveEndpointCodeTables` might be clearer as to what's in and out of scope of this function.

File: modules/GenericDataParser/src/server/generic_data_parser.R [hunk 2]
```diff
@@ -223,6 +223,77 @@ validateMetaData <- function(metaData, configList, username, formatSettings = li
...
+  # Saves incoming column data into their codetable / data dictionaries
+  # set up for endpoint manager data, only looking at "column name", "column units", "concentration units", "time units" for now... 
+  # We are going off just starting to just take one input columnName to save
```
Comment (bffrost) [inline modules/GenericDataParser/src/server/generic_data_parser.R:231]: nit: Is this comment line still relevant?

File: modules/GenericDataParser/src/server/generic_data_parser.R [hunk 3]
```diff
@@ -223,6 +223,77 @@ validateMetaData <- function(metaData, configList, username, formatSettings = li
...
+  ddictValues <- rbindlist(lapply(ddictLists, jsonlite::fromJSON))
+
+  # if the object is empty, we assume that no codes exist yet, so we will save all the input codes
+  if(nrow(ddictValues) == 0) {
+    codesToSave <- codes
+
+  #if the object is not empty, we will check to see which of the existing codes overlap with our input codes
+  } else {
+    #extract all the codes for a given codeKind
+    codesForCodeKind = ddictValues %>% filter(lsKind == codeKind) %>% pull(codeName)
```
Comment (bffrost) [inline modules/GenericDataParser/src/server/generic_data_parser.R:265]: I'm not sure `codeName` is the right attribute to pull here. If I'm understanding the flow so far,
codeName will be something like 'DDICT-00001" whereas the `shortName` attribute will be the relevant string containing the column name or units. Could you double check this?

You'll also be able to test this by uploading the same column name in two experiments, and seeing if you get duplicate entries in the Column Name dropdown. You can also confirm in the `ddict_value` table

File: modules/GenericDataParser/src/server/generic_data_parser.R [hunk 4]
```diff
@@ -3421,6 +3510,18 @@ runMain <- function(pathToGenericDataFormatExcelFile, reportFilePath=NULL,
       addError(paste0("Caught an error trying to validate dose response data. Please contact your administrator.: ", e$message))
     })
   }
+
+  # Validate Experiment Columns against Protocol Endpoints
+  # only proceed if it associated with an existing protocol (the protocol is not new)
+  if (!newProtocol) {
+    #if the protocol has strict endpoint matching enabled, proceed with validation
+    if (getProtocolStrictEndpointMatching(protocol) == TRUE) {
```
Comment (bffrost) [inline modules/GenericDataParser/src/server/generic_data_parser.R:3518]: We should do this validation either way. But the "strict" endpoint matching value should determine whether the validation throws errors (if strict) or warnings (if not strict).

It might make sense to pass the strictEndpointMatching value into `validateExperimentColumns` as an input variable.

However I think it would be a good idea to check the `client.protocol.endpointManager.enabled` config and skip this whole validation if that is set to false.

File: modules/GenericDataParser/src/server/generic_data_parser.R [hunk 5]
```diff
@@ -3442,9 +3543,20 @@ runMain <- function(pathToGenericDataFormatExcelFile, reportFilePath=NULL,
...
   }
   
+  # Save endpoint data codes
+  saveIncomingEndpointData(selColumnOrderInfo$Units, "column units")
+  saveIncomingEndpointData(selColumnOrderInfo$valueKind, "column name")
+  saveIncomingEndpointData(selColumnOrderInfo$concUnits, "concentration units")
+  saveIncomingEndpointData(selColumnOrderInfo$timeUnit, "time units")
+
+  ## SEL column order info
+  #saveSession('insideselColumnOrderInfo.RData')
```
Comment (bffrost) [inline modules/GenericDataParser/src/server/generic_data_parser.R:3553]: nit: you can remove this now since you're done debugging

File: modules/GenericDataParser/src/server/generic_data_parser.R [hunk 6]
```diff
@@ -3953,6 +4063,133 @@ getFormatParameters <- function(rawOnlyFormat, customFormatSettings, inputFormat
...
+          } else {
+            return(FALSE)
+          }   
+        }
+      }
+    }
+  }
+
+  # TODO - remove this after database migration
+  return(FALSE)
```
Comment (bffrost) [inline modules/GenericDataParser/src/server/generic_data_parser.R:4190]: I think it'd make more sense to return `racas::applicationSettings$client.protocol.strictEndpointMatchingDefault` if the protocol doesn't have this value set. It is the default after all.

File: modules/GenericDataParser/src/server/generic_data_parser.R [hunk 7]
```diff
@@ -3953,6 +4063,133 @@ getFormatParameters <- function(rawOnlyFormat, customFormatSettings, inputFormat
...
+      #if the experiment column matches one of the endpoints for the protocol, record it and move onto the next column
+      if (rowNamesMatch & rowUnitsMatch & rowTypesMatch) {
+        experimentRowMatchesEndpoint = TRUE
+        break 
+      }
+
+    }
+
+    if (experimentRowMatchesEndpoint == FALSE) {
+      addError(paste0("The result type '", experimentRowName, "' with data type '", experimentRowDataType, "' and units '", experimentRowUnits, "' is not one of the allowed result types for this protocol. Please revise your file or contact an ACAS administrator to update the allowed result types for this protocol."))
```
Comment (bffrost) [inline modules/GenericDataParser/src/server/generic_data_parser.R:4162]: Here's where you'll want to split between error & warning based on the "strict"ness of the protocol.

See the last image on https://confluence.schrodinger.com/x/egSgBw for some proposed verbiage.

File: modules/GenericDataParser/src/server/generic_data_parser.R [hunk 8]
```diff
@@ -3953,6 +4063,133 @@ getFormatParameters <- function(rawOnlyFormat, customFormatSettings, inputFormat
...
+  # Checks if the experiment columns names/units/data type are found in the protocol endpoint data
+  # Raises errors if an experiment column is not valid 
+  # Only used for protocols with strict endpoint enabled
+
+  for (experimentRowNum in seq(1, nrow(selColumnOrderInfo))) {
+    experimentRowData <- selColumnOrderInfo[experimentRowNum,]
+    experimentRowName = experimentRowData$valueKind
+    experimentRowUnits = experimentRowData$Units
+    experimentRowDataType = experimentRowData$valueType
+
```
Comment (bffrost) [inline modules/GenericDataParser/src/server/generic_data_parser.R:4121]: We should check concentration and conc units too

**Review by bffrost — CHANGES_REQUESTED** [review]
@hhan-schrodinger this is in pretty good shape. The code is well structured and easy to follow, especially with your code comments. I found a few areas that differ from the design but they all look like easy fixes.

**Review by dalejerikson — COMMENTED** [review]
I can't really comment in the context of the user story. However, I do really like your styling and comments. I think it's really clean and me actually not mind reading R (which is hard to do haha).

File: modules/GenericDataParser/src/server/generic_data_parser.R (hunk 1 above)
Comment (hhan-schrodinger) [reply modules/GenericDataParser/src/server/generic_data_parser.R:228]: Makes sense!

File: modules/GenericDataParser/src/server/generic_data_parser.R (hunk 2 above)
Comment (hhan-schrodinger) [reply modules/GenericDataParser/src/server/generic_data_parser.R:231]: Good catch!

File: modules/GenericDataParser/src/server/generic_data_parser.R (hunk 8 above)
Comment (hhan-schrodinger) [reply modules/GenericDataParser/src/server/generic_data_parser.R:4121]: Should we have additional logic to only validate if conc/conc units are enabled in the conf?

Comment (bffrost) [reply modules/GenericDataParser/src/server/generic_data_parser.R:4129]: That's a great question. Yes, we should add some logic for that case. ACAS is configured to not show concentration in the UI, and someone attempts to upload data that has concentrations into a strict match protocol, we should at the very least warn them. Something along the lines of:

```
The column you're uploading {column name} at concentration {conc} {conc units} cannot be fully validated against protocol {protocol name} because this ACAS server is not configured to track concentration in the Protocol Endpoints. To enable complete checking including concentration, please contact your administrator and ask them to enable "client.protocol.endpointManager.showConcentration"
```

File: modules/GenericDataParser/src/server/generic_data_parser.R (hunk 3 above)
Comment (hhan-schrodinger) [reply modules/GenericDataParser/src/server/generic_data_parser.R:263]: You are completely right! I think I misread the table. 

Screen Shot 2022-11-16 at 12 58 39 PM

File: modules/GenericDataParser/src/server/generic_data_parser.R [hunk 9]
```diff
@@ -3953,6 +4063,133 @@ getFormatParameters <- function(rawOnlyFormat, customFormatSettings, inputFormat
...
+}
+
+getProtocolStrictEndpointMatching <- function(protocol) {
+  # Check whether the protocol has strict endpoint matching set or not 
+  # Returns true or false
+
+  # TODO - This may cause errors in existing protocols that do not have strict endpoint matching already set
+  # To prevent those errors, we will automatically return false for now. 
+
+  for (lsState in protocol$lsStates) {
```
[the remark this answers is not in the document]
Comment (hhan-schrodinger) [reply modules/GenericDataParser/src/server/generic_data_parser.R]: Thanks, thats super helpful!

File: modules/GenericDataParser/src/server/generic_data_parser.R [hunk 10]
```diff
@@ -4064,13 +4058,16 @@ getFormatParameters <- function(rawOnlyFormat, customFormatSettings, inputFormat
...
   # Collects all the endpoint values in the protocol object and returns a data frame of them
 
   # First, we create an empty dataframe for the endpoint data associated with the protocol 
   protocolEndpointDataFrame <- data.frame(
     columnName=character(),
     units=character(),
-    dataType=character()
+    dataType=character(),
+    conc = character(),
+    concUnits = character()
```
Comment (bffrost) [inline modules/GenericDataParser/src/server/generic_data_parser.R:4070]: Are we missing Time & Time Units here?

**Review by bffrost — APPROVED** [review]
LGTM. I think I see a few places where we're not validating Time, but I'm comfortable with spinning that off as a separate case so you can get this PR merged. Nice work!

File: modules/GenericDataParser/src/server/generic_data_parser.R (hunk 10 above)
Comment (hhan-schrodinger) [reply modules/GenericDataParser/src/server/generic_data_parser.R:4070]: I meant to ask - I wasn't sure whether or not to include Time/Time Units initially. Can definitely add them here and the rest of the logic!
